packages/workspace-utils/src/index.ts

```diff
@@ -115,6 +116,54 @@ export async function getWorkspaceGradedFiles(
   return files;
 }
 
+/**
+ * Updates the disk usage of a workspace. This is computed as the sum of the
+ * sizes of all versions of the workspace. The result is stored in the `disk_usage`
+ * column of the `workspaces` table. The total size is returned.
+ *
+ * @param workspace_id The ID of the workspace to update.
+ * @param workspacesRoot The root directory of all workspace data.
+ */
+export async function updateWorkspaceDiskUsage(
+  workspace_id: string,
+  workspacesRoot: string,
+): Promise<number> {
+  const result = await queryOneRowAsync(sql.select_workspace, { workspace_id });
+  const workspace = result.rows[0];
+
+  // We'll compute the size for all versions of the workspace so that we don't need
+  // to separately store the size for each version.
+  const version = Number.parseInt(workspace.version, 10);
+
+  let totalSize = 0;
+  for (let i = 1; i <= version; i++) {
+    const workspaceVersionPath = path.join(
+      workspacesRoot,
+      `workspace-${workspace.id}-${workspace.version}`,
+    );
+    const size = await getDirectoryDiskUsage(workspaceVersionPath);
+    totalSize += size ?? 0;
+  }
+
+  await queryAsync(sql.update_workspace_disk_usage, { workspace_id, disk_usage: totalSize });
+
+  return totalSize;
+}
+
+async function getDirectoryDiskUsage(dir: string): Promise<number | null> {
+  try {
+    // macOS `du` doesn't support the `-b` flag, so we use `-k` instead and
+    // multiply the result by 1024 to convert from kilobytes to bytes.
+    const res = await execa('du', ['-sk', dir]);
+    const [size] = res.stdout.split(/\s/);
+    return Number.parseInt(size, 10) * 1024;
```

Comment:
An alternative to this would be to implement this in pure JS; this would not be difficult and would give us slightly more accurate results and avoids potential platform-specific issues with `du`.

Reply:
For reference, I did go off and implement it in JS, it's not bad at all:

```js
const fs = require('node:fs/promises');
const path = require('node:path');

async function totalSize(rootPath) {
  let size = 0;
  const stack = [rootPath];

  while (stack.length) {
    const currentPath = stack.pop();
    const stats = await fs.stat(currentPath);

    size += stats.size;

    if (stats.isDirectory()) {
      const files = await fs.readdir(currentPath);
      for (const file of files) {
        stack.push(path.join(currentPath, file));
      }
    }
  }

  return size;
}
```

Reply:
This includes files in other file systems. Do you want that?

Reply:
I'm not sure what you mean by that?

Reply:
I can do a proper review next week, but have you considered using `df` instead of `du`? It takes the information from the disks themselves, which should be faster.

Reply:
Indeed, I did not consider volume mounts, my mistake. Anyway, the mountpoint only includes the "home" directory, not any peripheral information that is stored outside of home (e.g., /tmp), that is initially retrieved from the image but can then change with the use of the workspace. My understanding is that this data persists, is that correct? Do you want to include that in the calculation of workspace size?

In any case: you should use `fs.lstat()` instead of `fs.stat()` to avoid problems with symlinks (`fs.lstat()` will return the stats of the symbolic link, `stat()` will return the stats of the file it points to). If a student has a symlink in the home directory that points to a large file outside of home, the large file should not be counted towards this limit; if a student has a symlink to another directory, the directory should not be included in the calculation.

Also, you can use `fs.readdir(rootPath, { recursive: true })` to avoid the need for a stack.

If you want to use disk space instead of payload size, you can use `stats.blocks * stats.blksize` instead of `stats.size`, though the difference will likely be relatively small in the cases that matter to you.

Reply:
> Anyway, the mountpoint only includes the "home" directory, not any peripheral information that is stored outside of home (e.g., /tmp), that is initially retrieved from the image but can then change with the use of the workspace. My understanding is that this data persists, is that correct? Do you want to include that in the calculation of workspace size?

The data does not persist. The _only_ data that persists when a workspace is stopped is what is mounted in to the home directory.

> If a student has a symlink in the home directory that points to a large file outside of home, the large file should not be counted towards this limit; if a student has a symlink to another directory, the directory should not be included in the calculation.

That's a good reminder to consider symlinks. However, as we'll be running `df` (or whatever method we choose), from outside the container, we actually shouldn't consider symlinks if they resolve to a file outside the workspace home directory.

> Also, you can use `fs.readdir(rootPath, { recursive: true })` to avoid the need for a stack.

This is a great idea, I'll do this.

> If you want to use disk space instead of payload size, you can use `stats.blocks * stats.blksize` instead of `stats.size`, though the difference will likely be relatively small in the cases that matter to you.

I like this in principle, and it does more closely match how AWS EFS measures files: https://docs.aws.amazon.com/efs/latest/ug/metered-sizes.html. However, I've observed that Node's own `fs.stat` reports `blksize` as `1048576` (1MB), even though `stat --format "%b*%B" ...` shows that the block size is 512. I imagine that Node might be using either `rsize` or `wsize` (the options we use when mounting the EFS volume to the host).  Using Node's version of the block size would grossly inflate the actual size. I think `du -sk` is actually going to give us the best possible value?

That said, I'm confused by the behavior of `du` on macOS. e.g.:

```sh
du -cak .github
4       .github/workflows/ci.yml
4       .github/workflows
4       .github/dependabot.yml
8       .github
8       total
```

It's almost as if it doesn't count the blocks used by the directories themselves? Not that this really matters for production usage, of course.

Reply:
I think macOS has a BSD version of du that might behave differently. Can you install GNU coreutils with brew or something and use their version of du for consistency?

Reply:
`gdu -cak` (the `g` prefix means it's the version from GNU coreutils) prints the same thing.

Reply:
Oh, maybe something different about APFS then.

Reply:
> That's a good reminder to consider symlinks. However, as we'll be running `df` (or whatever method we choose), from outside the container, we actually shouldn't consider symlinks if they resolve to a file outside the workspace home directory.

To clarify, we should consider symlinks, we shouldn't follow them (regardless of where we run this command). So if there is a symlink, it should be treated and accounted for as a symlink. The way to do that (if you're checking stats) is to use `fs.lstat` instead of `fs.stat`.

Reply:
Yep, that's what I had in mind, I just didn't say it as precisely as you.

Reply:
Yes, as long as you don't use the -L or -D options.

```
       -D, --dereference-args
              dereference only symlinks that are listed on the command line

       -H     equivalent to --dereference-args (-D)

       -L, --dereference
              dereference all symbolic links

       -P, --no-dereference
              don't follow any symbolic links (this is the default)
```

`du` also has the advantage that if two files share the same inode (i.e., are hard links), they only count once.